employee-rostering-frontend/src/ui/components/DataTable.tsx

```diff
@@ -33,50 +33,45 @@ export interface DataTableProps<T> {
 }
 
 interface DataTableState<T> {
-  newRow: IRow | null;
+  newRowData: T|null;
```

Comment:
This is confusing. The state property is being used for two different purposes:

1. As a flag indicating that `Add` button has been clicked and a new entry row should be rendered (correct).

2. To hold the new data. This would be correct if the data were updated using `setState()` but you're allowing a child component (e.g. a `TextInput`) to change (mutate) the data directly without going through `setState()` (see [Using State Correctly](https://reactjs.org/docs/state-and-lifecycle.html#using-state-correctly). It works. I don't deny that. But mixing the responsibilities increases mental capacity required to understand what's going on. And this approach is also against [The Data Flows Down](https://reactjs.org/docs/state-and-lifecycle.html#the-data-flows-down) principle.

    Instead, DataTable should pass the TextInput an onChange() handler function that updates the state with the new value. It can also validate it and set `isValid` flag in the state, which might be a prop of the Save button and you'll get as-you-type validation.